src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,187 @@
+                }
+            }
+            throw ADP.Argument(SR.ADP_DbProviderFactories_NoAssemblyQualifiedName);
+        }
+
+        public static DataTable GetFactoryClasses()
+        {
+            try
+            {
+                _providerTableLock.EnterReadLock();
+                return _providerTable == null ? GetInitialProviderTable() : _providerTable.Copy();
+            }
+            finally
+            {
+                _providerTableLock.ExitReadLock();
+            }
+        }
+
+        public static DbProviderFactory GetFactory(DbConnection connection)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            return connection.ProviderFactory;
+        }
+
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName)
+        {
+            ConfigureFactory(providerFactoryClass, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(providerFactoryClass, nameof(providerFactoryClass));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            if (!providerFactoryClass.IsSubclassOf(typeof(DbProviderFactory)))
+            {
+                throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_NotAFactoryType, providerFactoryClass.FullName));
+            }
+            RegisterFactoryInTable(providerFactoryClass, providerInvariantName, name, description);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName)
+        {
+            ConfigureFactory(connection, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            DbProviderFactory factoryInstance = GetFactory(connection);
+            if (factoryInstance == null)
+            {
+                throw ADP.Argument(SR.ADP_DbProviderFactories_NoFactorySuppliedThroughDbConnection);
+            }
+            RegisterFactoryInTable(factoryInstance.GetType(), providerInvariantName, name, description);
+        }
+        
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
+            string invariantNameToUse = string.IsNullOrWhiteSpace(providerInvariantName) ? factoryType.Namespace : providerInvariantName;
```

Comment:
It seems in the two places where we call this private method we don't allow `providerInfariantName` to be null. Do we still need this fallback to the namespace?